### ISSUE TITLE
Bug: Crash when destroying all associations

### ISSUE THREAD

**Issue by jamesst20** [brendon/positioning#16]
**Exemple**

```ruby
class PostType < ApplicationRecord
    positioned column: :position

    belongs_to :template, class_name: "Rcf::Field"

    has_many :post_groups, dependent: :destroy

    before_save :destroy_post_groups_on_template_change
 
   private

   def destroy_post_groups_on_template_change
     return unless will_save_change_to_template_id?

     PostGroup.where(post_type_id: template_id_was).destroy_all
   end
```

<img width="1078" alt="Capture d’écran, le 2024-07-08 à 23 17 29" src="https://github.com/brendon/positioning/assets/490234/39ce59aa-55c8-4119-887b-778b65c0a290">

**Stacktrace**

```
web    | NoMethodError (undefined method `+' for nil):
web    |
web    | positioning (0.2.3) lib/positioning/mechanisms.rb:44:in `destroy_position'
web    | positioning (0.2.3) lib/positioning.rb:56:in `block in positioned'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:448:in `instance_exec'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:448:in `block in make_lambda'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:202:in `block (2 levels) in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:707:in `block (2 levels) in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `catch'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `block in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:203:in `block in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `block in invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `each'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:109:in `run_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:952:in `_run_destroy_callbacks'
web    | activerecord (7.1.3.4) lib/active_record/callbacks.rb:423:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `block in destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:365:in `block in with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:361:in `with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/persistence.rb:797:in `destroy!'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_many_association.rb:129:in `each'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_many_association.rb:129:in `delete_records'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:396:in `remove_records'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:387:in `block in delete_or_destroy'
web    | activerecord (7.1.3.4) lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:212:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:314:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:387:in `delete_or_destroy'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:194:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:171:in `destroy_all'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_many_association.rb:29:in `handle_dependency'
web    | activerecord (7.1.3.4) lib/active_record/associations/builder/association.rb:141:in `block in add_destroy_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:470:in `instance_exec'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:470:in `block in make_lambda'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:202:in `block (2 levels) in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:707:in `block (2 levels) in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `catch'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `block in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:203:in `block in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `block in invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `each'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:109:in `run_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:952:in `_run_destroy_callbacks'
web    | activerecord (7.1.3.4) lib/active_record/callbacks.rb:423:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `block in destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:365:in `block in with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:361:in `with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_one_association.rb:33:in `delete'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_one_association.rb:22:in `handle_dependency'
web    | activerecord (7.1.3.4) lib/active_record/associations/builder/association.rb:141:in `block in add_destroy_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:470:in `instance_exec'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:470:in `block in make_lambda'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:202:in `block (2 levels) in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:707:in `block (2 levels) in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `catch'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `block in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:203:in `block in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `block in invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `each'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:109:in `run_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:952:in `_run_destroy_callbacks'
web    | activerecord (7.1.3.4) lib/active_record/callbacks.rb:423:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `block in destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:365:in `block in with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:361:in `with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/persistence.rb:797:in `destroy!'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_many_association.rb:129:in `each'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_many_association.rb:129:in `delete_records'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:396:in `remove_records'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:387:in `block in delete_or_destroy'
web    | activerecord (7.1.3.4) lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:212:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:314:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:387:in `delete_or_destroy'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:194:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/associations/collection_association.rb:171:in `destroy_all'
web    | activerecord (7.1.3.4) lib/active_record/associations/has_many_association.rb:29:in `handle_dependency'
web    | activerecord (7.1.3.4) lib/active_record/associations/builder/association.rb:141:in `block in add_destroy_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:470:in `instance_exec'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:470:in `block in make_lambda'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:202:in `block (2 levels) in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:707:in `block (2 levels) in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `catch'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:706:in `block in default_terminator'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:203:in `block in halting'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `block in invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `each'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:598:in `invoke_before'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:109:in `run_callbacks'
web    | activesupport (7.1.3.4) lib/active_support/callbacks.rb:952:in `_run_destroy_callbacks'
web    | activerecord (7.1.3.4) lib/active_record/callbacks.rb:423:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `block in destroy'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:365:in `block in with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:361:in `with_transaction_returning_status'
web    | activerecord (7.1.3.4) lib/active_record/transactions.rb:305:in `destroy'
web    | activerecord (7.1.3.4) lib/active_record/relation.rb:625:in `each'
web    | activerecord (7.1.3.4) lib/active_record/relation.rb:625:in `destroy_all'
web    | /Users/james/Projects/rcf/app/models/rcf/post_type.rb:51:in `destroy_post_groups_on_template_change'
...
```

**Comment by brendon** [comment]
Hi @jamesst20, that's a confusing one. It might help to get you to isolate this out into a reproducible test in our test suite. Are you able to reproduce it in a single test that way and do a PR with that in it and then I can look into a solution?

Just checking you don't have a default set for any of your position columns at the database level?

This also just generally rings a bell, can you confirm you're using the latest version of the gem as I think someone else had a problem with `destroy_all` and we fixed it by changing the destruction process.

**Comment by jamesst20** [comment]
> Hi @jamesst20, that's a confusing one. It might help to get you to isolate this out into a reproducible test in our test suite. Are you able to reproduce it in a single test that way and do a PR with that in it and then I can look into a solution?
> 
> Just checking you don't have a default set for any of your position columns at the database level?
> 
> This also just generally rings a bell, can you confirm you're using the latest version of the gem as I think someone else had a problem with `destroy_all` and we fixed it by changing the destruction process.

Hi Brendon,

I indeed don't have a default in my columns. I do have unique non-null constraint if I am not mistaken.

I am using the latest version of the gem, the stacktrace also gives the version (0.2.3).

I will check for a reproductible exemple

**Comment by brendon** [comment]
Thanks @jamesst20, good to know. I'm very happy to help, but it'll be definitely easier if I can have an example I can play with on my end :)

**Comment by jamesst20** [comment]
> Thanks @jamesst20, good to know. I'm very happy to help, but it'll be definitely easier if I can have an example I can play with on my end :)

Hey @brendon ! It took me a while to finally be able to reproduce it in an isolated demo! The reason it was so hard to reproduce is that it seems to only happen when using the gem "Goldiloader".

I am not sure why there is a difference in behavior there, but I hope this can help you figure it out! :)

```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'rails'
  gem 'sqlite3', '~> 1.4'
  gem 'positioning'
  gem 'goldiloader'
end

require 'active_record'
require 'positioning'
require 'minitest/autorun'
require 'goldiloader'
require 'logger'

Goldiloader.globally_enabled = true

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new($stdout)

ActiveRecord::Schema.define do
  create_table 'fields', force: :cascade do |t|
    t.bigint 'parent_id'

    t.string :name

    t.integer :position
    t.index %i[parent_id position], unique: true
  end

  create_table 'posts', force: :cascade do |t|
    t.string :name
  end

  create_table 'post_values', force: :cascade do |t|
    t.bigint 'parent_id'
    t.references :post, null: true
    t.references :field, null: false

    t.string :value

    t.integer :position
    t.index %i[parent_id position], unique: true
  end
end

class Field < ActiveRecord::Base
  positioned on: :parent, column: :position

  belongs_to :parent, class_name: 'Field', inverse_of: :subfields, optional: true

  has_many :subfields, foreign_key: :parent_id, class_name: 'Field', dependent: :destroy
  has_many :post_values, dependent: :destroy, inverse_of: :parent

  accepts_nested_attributes_for :post_values, allow_destroy: true, reject_if: :all_blank
end

class Post < ActiveRecord::Base
  has_one :post_value, dependent: :destroy

  accepts_nested_attributes_for :post_value, allow_destroy: true, reject_if: :all_blank
end

class PostValue < ActiveRecord::Base
  positioned on: :parent, column: :position

  belongs_to :field, inverse_of: :post_values
  belongs_to :post, inverse_of: :post_value, optional: true
  belongs_to :parent, class_name: 'PostValue', inverse_of: :subvalues, optional: true

  has_many :subvalues,
           -> { order(position: :asc) },
           foreign_key: :parent_id,
           class_name: 'PostValue',
           dependent: :destroy,
           inverse_of: :parent

  accepts_nested_attributes_for :subvalues, allow_destroy: true, reject_if: :all_blank
end

class BugTest < Minitest::Test
  def post
    @post ||= Post.create!(name: 'Post 1')
  end

  def field
    @field ||= Field.create!(
      name: 'Field 1',
      position: 1,
      subfields: [
        Field.new(name: 'Field 1.1', subfields: [
                    Field.new(name: 'Field 1.1.1', subfields: [
                                Field.new(name: 'Field 1.1.1.1'),
                                Field.new(name: 'Field 1.1.1.2'),
                                Field.new(name: 'Field 1.1.1.3')
                              ]),
                    Field.new(name: 'Field 1.1.2', subfields: [
                                Field.new(name: 'Field 1.1.2.1'),
                                Field.new(name: 'Field 1.1.2.2'),
                                Field.new(name: 'Field 1.1.2.3')
                              ]),
                    Field.new(name: 'Field 1.1.3', subfields: [
                                Field.new(name: 'Field 1.1.3.1'),
                                Field.new(name: 'Field 1.1.3.2'),
                                Field.new(name: 'Field 1.1.3.3')
                              ])
                  ]),
        Field.new(name: 'Field 1.2', subfields: [
                    Field.new(name: 'Field 1.2.1', subfields: [
                                Field.new(name: 'Field 1.2.1.1'),
                                Field.new(name: 'Field 1.2.1.2'),
                                Field.new(name: 'Field 1.2.1.3')
                              ]),
                    Field.new(name: 'Field 1.2.2', subfields: [
                                Field.new(name: 'Field 1.2.2.1'),
                                Field.new(name: 'Field 1.2.2.2'),
                                Field.new(name: 'Field 1.2.2.3')
                              ]),
                    Field.new(name: 'Field 1.2.3', subfields: [
                                Field.new(name: 'Field 1.2.3.1'),
                                Field.new(name: 'Field 1.2.3.2'),
                                Field.new(name: 'Field 1.2.3.3')
                              ])
                  ])
      ]
    )
  end

  def post_values
    PostValue.create!(
      field: field,
      value: 'Value 1',
      subvalues: [
        PostValue.new(value: 'Value 1.1', field: field.subfields[0], subvalues: [
                        PostValue.new(value: 'Value 1.1.1', field: field.subfields[0].subfields[0], subvalues: [
                                        PostValue.new(value: 'Value 1.1.1.1',
                                                      field: field.subfields[0].subfields[0].subfields[0]),
                                        PostValue.new(value: 'Value 1.1.1.2',
                                                      field: field.subfields[0].subfields[0].subfields[1]),
                                        PostValue.new(value: 'Value 1.1.1.3',
                                                      field: field.subfields[0].subfields[0].subfields[2])
                                      ])
                      ]),
        PostValue.new(value: 'Value 1.2', field: field.subfields[0], subvalues: [
                        PostValue.new(value: 'Value 1.2.1', field: field.subfields[1].subfields[0], subvalues: [
                                        PostValue.new(value: 'Value 1.1.1.1',
                                                      field: field.subfields[1].subfields[0].subfields[0]),
                                        PostValue.new(value: 'Value 1.1.1.2',
                                                      field: field.subfields[1].subfields[0].subfields[1]),
                                        PostValue.new(value: 'Value 1.1.1.3',
                                                      field: field.subfields[1].subfields[0].subfields[2])
                                      ])
                      ])
      ]
    )
  end

  def setup
    post
    field
    post_values
  end

  def test_destroy_not_crash
    Field.find_by(parent_id: nil).destroy!
  end
end
```

**Comment by brendon** [comment]
Thanks @jamesst20, I think in this case it's a bit outside of what I can provide support for since it's another gem potentially interfering with our queries.

You've made a good start now though and I'm happy to apply any fixes you come up with that we need to make on this end. Are you happy to dig into your test case and see if you can identify what `goldiloader` is doing that breaks things? It could be that we need to be more explicit about something, or that `goldiloader` is the gem that needs fixing. To me it kind of seems like `goldiloader` is loading or preloading models where we're expecting them to be fresh when we fetch them.

**Comment by jamesst20** [comment]
> Thanks @jamesst20, I think in this case it's a bit outside of what I can provide support for since it's another gem potentially interfering with our queries.
> 
> You've made a good start now though and I'm happy to apply any fixes you come up with that we need to make on this end. Are you happy to dig into your test case and see if you can identify what `goldiloader` is doing that breaks things? It could be that we need to be more explicit about something, or that `goldiloader` is the gem that needs fixing. To me it kind of seems like `goldiloader` is loading or preloading models where we're expecting them to be fresh when we fetch them.

It's indeed possible the issue is not on`positionning` end, however Goldiloader is a pretty popular gem too so I believe both positioning and goldiloader is well tested. There is this gem https://github.com/DmitryTsepelev/ar_lazy_preload that is similar and the error is not present though.

I would like to help as much as possible but I really am not familiar with the internal of either gem.

I have added a breakpoint right before the crash happens in case this might make you curious:

![Capture d’écran, le 2024-07-09 à 21 37 14](https://github.com/brendon/positioning/assets/490234/cedafeae-f102-4d64-96ee-41b01c62a4e0)

**Comment by brendon** [comment]
Interesting, it's probably the case that you're relying on a cached collection of records?

You may want to explore the `uncached` method block to try pulling in the association without any query caching before running your `destroy_all`.

It might be worth trying to trigger our association destroy detection:

https://github.com/brendon/positioning/blob/0c3be35b233744d68720c6acf65dcad9308c9b7f/lib/positioning/mechanisms.rb#L188-L192

Normally if one of the scopes of a list item is also being deleted we skip all the gap filling logic when destroying the list item because all of the list items in that scope will be deleted anyway. You're manually destroying the whole scope's children so there might be a way to indicate that this can be skipped. `destroyed_by_association=(reflection)`

**Comment by jamesst20** [comment]
> Interesting, it's probably the case that you're relying on a cached collection of records?
> 
> You may want to explore the `uncached` method block to try pulling in the association without any query caching before running your `destroy_all`.
> 
> It might be worth trying to trigger our association destroy detection:
> 
> https://github.com/brendon/positioning/blob/0c3be35b233744d68720c6acf65dcad9308c9b7f/lib/positioning/mechanisms.rb#L188-L192
> 
> Normally if one of the scopes of a list item is also being deleted we skip all the gap filling logic when destroying the list item because all of the list items in that scope will be deleted anyway. You're manually destroying the whole scope's children so there might be a way to indicate that this can be skipped. `destroyed_by_association=(reflection)`

(Have you seed `edit #4`? I submitted almost at same time you replied)

I have given a try to your suggestion by completly disabling query cache
```
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:', query_cache: false)
```

and I also tried

```ruby
ActiveRecord::Base.uncached do
  Field.find_by(parent_id: nil).destroy!
end
```

but it didn't make a difference.

By looking at `destroyed_via_positioning_scope?` and seems it doesn't ever return true in my test case no matter if I use goldiloader or not

**Comment by jamesst20** [comment]
@brendon  

Thanks for your time! It's very much appreciated. After further investigation, I believe the bug is on goldiloader end.

I have updated my reproductible exemple for them so they can see easily some events are duplicated

**Comment by brendon** [comment]
No worries @jamesst20. Happy to be the rubber duck! :D

Well done on not including `positioning` in your test case in the new issue. That'll be appreciated by that maintainer :D

I'll close this for now but let me know if we need to look at anything again.

**Comment by jamesst20** [comment]
Hey @brendon 

Just doing a follow up on this! It seems that when we use preloading with ActiveRecord, it triggers callbacks multiple times. Callbacks such as `before_destroy`

It doesn't appear to be on goldiloader end specifically and I was able to reproduce with similar gem such as https://github.com/DmitryTsepelev/ar_lazy_preload and I also was able to reproduce without any gem by simply calling `includes(...)`

I'm linking the issue: https://github.com/rails/rails/issues/52714

**Comment by brendon** [comment]
Thanks @jamesst20, sounds like a truely bizarre problem! Surely it's got something to do with some kind of shared class variable or thread safety issues?

**Comment by donnykurnia** [comment]
Hi all,

I encountered this bug in my client's project. The project didn't use the `goldiloader` gem.
I noticed that this happened when calling the `remove_all` class method. I want to know if I change it into `delete_all` so the callback won't fire, will it affect the next time the `positioning` gem handles the leftover data? Or will it handle the 'hole' automatically?

**Comment by brendon** [comment]
Hi @donnykurnia, can you write me a failing test that simulates this issue? I can look at it from there :)

**Comment by donnykurnia** [comment]
Hi @brendon I've written tests to simulate these bugs. But I still haven't been able to trigger the error. In production, the exception is logged in Rollbar only once in a while.

I've looked up the related code. It is caused by the `nil` value of the `position_was` result. The code is using `.pick`, and there are only 3 possible ways that `pick` returns `nil`. I think this is caused by a race condition, where the previously deleted row then somehow gets processed again by the callback.

I don't have enough time to investigate this further. For my use case, using `delete_all` in my code avoids the callback entirely. Perhaps this can help the other affected dev out there.

**Comment by brendon** [comment]
No worries @donnykurnia, perhaps it something you can throw AI at and see what it can surface? Keep me posted if you find out any more.
